### ISSUE TITLE
wayfire-plugins-extra doesn't install in NixOS/Nixpkgs

### ISSUE THREAD

**Issue by AndersonTorres** [WayfireWM/wayfire-plugins-extra#78]
Hello, people!

I am Anderson Torres, a Nixpkgs developer. I am trying to package wayfire-plugins-extra to nixpkgs, but it is not working.

<details><summary>A small background about Nixpkgs:</summary>

  - Nix is a functional programming language tailored for package and systems management. Nix is also the name of the interpreter that receives Nix expressions and outputs software packages. 
    - Being a functional programming language, it treats packages in a peculiar manner, in a similar way Haskell treats values: packages are built via functions that don't have side effects. In particular, once a package is installed, it can't be modified. 
  In traditional package managers, all packages are scattered in a same root tree, `/`. On the other hand, Nix installs each package in a dedicated read-only directory inside `/nix/store`.
  An example: firefox is installed at
`/nix/store/gv4rgr1p1dq5s4hx7ald6a7kli6p3xrz-firefox-85.0/` (that hash part of the directory name encodes the dependency tree of Firefox).
This way, it is ensured that Firefox can' t be modified once it is installed.

- Nixpkgs is a (large!) set of Nix expressions that build many softwares, from compilers and assemblers to entire desktop environments like KDE and GNOME. 
- NixOS is a Linux distro built with a specifically tailored set of Nix expressions that build an entire OS environment, from bootloader and kernel to userland programs and desktop environments. Think on it as a large Nix expression that outputs an entire Linux distro.
</details>

TL;DR - Nixpkgs doesn' t allow modifications on an already installed package.

I think this is the reason why I can't install wayfire-plugins-extra. The last log lines show this:

<details> <summary>Installation output:</summary>

```

wayfire-plugins-extra> [0/1] Installing files.
wayfire-plugins-extra> Installing src/libannotate.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libautorotate-iio.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libbackground-view.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libbench.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libcrosshair.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libfollow-focus.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libforce-fullscreen.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libglib-main-loop.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libjoin-views.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libkeycolor.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libmag.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libshowrepaint.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libwater.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libwinzoom.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing src/libworkspace-names.so to /nix/store/2qc95nl93ws2ijyl936jmawg43w4m9jb-wayfire-plugins-extra-0.7.0/lib/wayfire
wayfire-plugins-extra> Installing /build/wayfire-plugins-extra-0.7.0/metadata/annotate.xml to /nix/store/4sc2544i02gx19w9wp26sjgh0fbr8mp8-wayfire-0.7.0/share/wayfire/metadata
wayfire-plugins-extra> Traceback (most recent call last):
wayfire-plugins-extra>   File "/nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/mesonmain.py", line 140, in run
wayfire-plugins-extra>     return options.run_func(options)
wayfire-plugins-extra>   File "/nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 554, in run
wayfire-plugins-extra>     installer.do_install(datafilename)
wayfire-plugins-extra>   File "/nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 375, in do_install
wayfire-plugins-extra>     self.install_data(d)
wayfire-plugins-extra>   File "/nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 406, in install_data
wayfire-plugins-extra>     if self.do_copyfile(fullfilename, outfilename, makedirs=(d.dirmaker, outdir)):
wayfire-plugins-extra>   File "/nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/lib/python3.8/site-packages/mesonbuild/minstall.py", line 278, in do_copyfile
wayfire-plugins-extra>     shutil.copy2(from_file, to_file)
wayfire-plugins-extra>   File "/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/lib/python3.8/shutil.py", line 435, in copy2
wayfire-plugins-extra>     copyfile(src, dst, follow_symlinks=follow_symlinks)
wayfire-plugins-extra>   File "/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/lib/python3.8/shutil.py", line 264, in copyfile
wayfire-plugins-extra>     with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
wayfire-plugins-extra> PermissionError: [Errno 13] Permission denied: '/nix/store/4sc2544i02gx19w9wp26sjgh0fbr8mp8-wayfire-0.7.0/share/wayfire/metadata/annotate.xml'
wayfire-plugins-extra> FAILED: meson-install
wayfire-plugins-extra> /nix/store/sm3v40fwgs85f23vq250wdfl6l54xkm2-meson-0.56.0/bin/meson install --no-rebuild
wayfire-plugins-extra> ninja: build stopped: subcommand failed.
error: --- Error ----------------------------------------------------------- nix
builder for '/nix/store/rz8v2vfxp7snsbiz2wci03lqk2ajxymk-wayfire-plugins-extra-0.7.0.drv' failed with exit code 1

```

</details>

Highlight the line below:

```
Installing /build/wayfire-plugins-extra-0.7.0/metadata/annotate.xml to /nix/store/4sc2544i02gx19w9wp26sjgh0fbr8mp8-wayfire-0.7.0/share/wayfire/metadata
```

This line causes an error, because, as I said, Nixpkgs doesn't allow modification of an already installed package (namely, wayfire).

P.S.: this is the offending file:

https://github.com/WayfireWM/wayfire-plugins-extra/blob/5c4ed0d593c08e23869dc9f1a125f26d883d14eb/metadata/meson.build#L1

I have zero experience with Meson (indeed I am studying it right now), but these lines clearly try to install tfiles in specific wayfire directories, and it violates the separation enforced by Nix.

**Comment by AdrianVovk** [comment]
Hello! I'm familiar with Nix

Wayfire Plugins install files to the standard search location, /usr/share/wayfire/metadata path. Wayfire searches this location by default

You can override this with an environment variables (check the wiki). So basically, you're going to want to wrap wayfire to set the env variable to a location you can assemble files into and then assemble all the packages that have a /usr/share/wayfire into this location

**Comment by soreau** [comment]
Closing this since this is an issue with the way nix works, not wayfire-plugins-extra build system.

**Comment by ztmr** [comment]
@AndersonTorres did you get anywhere with this?

**Comment by Electrostasy** [comment]
> @AndersonTorres did you get anywhere with this?

I packaged wayfire-plugins-extra [here](https://github.com/Electrostasy/dots/blob/c74615746d326142058df6e4813ab823138c4b73/packages/wayfire/wayfirePlugins/wayfire-plugins-extra/default.nix). I found that `libdir` and `metadatadir` have to be overwritten via `PKG_CONFIG_$PACKAGE_$VARIABLE` environment variables so that meson doesn't try to install to the Wayfire Nix store path.
